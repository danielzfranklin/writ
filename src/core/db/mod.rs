pub mod author;
pub mod blob;
pub mod cache;
pub mod commit;
pub mod object;
pub mod tree;

pub use author::Author;
pub use blob::Blob;
pub use commit::Commit;
pub use object::{Object, ObjectBuilder, Oid, UntypedOid};
pub use tree::Tree;

use bstr::{BString, ByteSlice};
use flate2::{read::ZlibDecoder, write::ZlibEncoder, Compression};
use tempfile::NamedTempFile;

use std::{
    collections::BTreeMap,
    fs::{self, File},
    io::{self, BufRead, BufReader, BufWriter, ErrorKind, Read, Write},
    num::ParseIntError,
    path::PathBuf,
};

use self::cache::Cache;
use crate::core::WsPath;

/// Note: Cloning doesn't keep the cache
#[derive(Debug)]
pub struct Db {
    path: PathBuf,
    cache: Cache,
}

impl Db {
    pub fn new<P: Into<PathBuf>>(git_dir: P) -> Self {
        Self {
            path: git_dir.into().join("objects"),
            cache: Cache::new(),
        }
    }

    pub fn load<O: Object>(&mut self, oid: Oid<O>) -> Result<O, LoadError<O>> {
        if let Some(cached) = self.cache.get(&oid) {
            return Ok(cached.clone());
        }

        let (len, bytes) = self.load_bytes(O::TYPE, &oid)?;
        let object = O::deserialize(oid, len, bytes).map_err(|e| LoadError::Deserialize(oid, e))?;
        self.cache.insert(oid, object.clone());

        Ok(object)
    }

    pub fn load_tree_file(
        &mut self,
        mut tree: Oid<Tree>,
        path: &WsPath,
    ) -> Result<Option<tree::FileNode>, LoadError<Tree>> {
        for parent in path.parents() {
            if let Some(tree::Node::Tree { oid: next_tree, .. }) =
                self.load(tree)?.direct_child(parent.file_name())
            {
                tree = *next_tree;
            } else {
                return Ok(None);
            }
        }

        if let Some(tree::Node::File(file)) = self.load(tree)?.direct_child(path.file_name()) {
            Ok(Some(file.clone()))
        } else {
            Ok(None)
        }
    }

    pub fn load_tree_files(
        &mut self,
        root: &WsPath,
        tree: Oid<Tree>,
    ) -> Result<BTreeMap<WsPath, tree::FileNode>, LoadError<Tree>> {
        let mut files = BTreeMap::new();
        self.load_all_tree_files_into(&mut files, root, tree)?;
        Ok(files)
    }

    fn load_all_tree_files_into(
        &mut self,
        files: &mut BTreeMap<WsPath, tree::FileNode>,
        root: &WsPath,
        tree: Oid<Tree>,
    ) -> Result<(), LoadError<Tree>> {
        let tree = self.load(tree)?;
        for child in tree.direct_children() {
            match child {
                tree::Node::File(file) => {
                    let path = root.join_bytes(child.name());
                    files.insert(path, file.clone());
                }
                tree::Node::Tree { oid, .. } => {
                    let next_root = root.join_bytes(child.name());
                    self.load_all_tree_files_into(files, &next_root, *oid)?;
                }
            }
        }
        Ok(())
    }

    /// Doesn't cache
    fn load_bytes<O: Object>(
        &self,
        expected_type: &[u8],
        oid: &Oid<O>,
    ) -> Result<(usize, impl BufRead), LoadBytesError<O>> {
        let path = self.oid_path(&oid);
        let file = match File::open(&path) {
            Ok(file) => Ok(file),
            Err(err) if err.kind() == io::ErrorKind::NotFound => {
                Err(LoadBytesError::NotFound(*oid))
            }
            Err(err) => Err(LoadBytesError::Open(*oid, err)),
        }?;

        let mut bytes = BufReader::new(ZlibDecoder::new(file));

        let mut o_type = vec![0; expected_type.len() + 1];
        bytes
            .read_exact(&mut o_type)
            .map_err(|e| LoadBytesError::ReadPrefix(*oid, e))?;
        let sep = o_type.pop().unwrap();
        if sep != b' ' {
            return Err(LoadBytesError::Corrupt(*oid));
        }
        if o_type != expected_type {
            return Err(LoadBytesError::WrongType {
                oid: *oid,
                expected: expected_type.into(),
                actual: o_type.into(),
            });
        }

        let mut len = BString::from(Vec::new());
        bytes
            .read_until(b'\0', &mut len)
            .map_err(|e| LoadBytesError::ReadPrefix(*oid, e))?;
        len.pop().unwrap();
        let len = len
            .to_str()
            .map_err(|e| LoadBytesError::ParseLenToBytes(*oid, e))?;
        let len: usize = len
            .parse()
            .map_err(|e| LoadBytesError::ParseLenToInt(*oid, e))?;

        Ok((len, bytes))
    }

    /// Doesn't cache
    pub fn store_bytes<OB: ObjectBuilder>(&self, content: &[u8]) -> StoreResult<OB::Object> {
        fn helper<O: Object>(db: &Db, oid: &Oid<O>, bytes: &[u8]) -> io::Result<()> {
            let path = db.oid_path(oid);

            if path.exists() {
                return Ok(());
            }

            let mut temp = NamedTempFile::new()?;

            {
                let mut writer = BufWriter::new(&mut temp);
                let mut writer = ZlibEncoder::new(&mut writer, Compression::default());
                writer.write_all(bytes)?;
            }

            // We use a temp file to get an atomic write
            temp.flush()?;

            match fs::rename(temp.path(), &path) {
                Err(err) if err.kind() == ErrorKind::NotFound => {
                    fs::create_dir(&path.parent().expect("has parent"))?;
                    fs::rename(temp.path(), &path)?;
                }
                Err(err) => return Err(err),
                Ok(()) => (),
            }

            Ok(())
        }

        let o_type = OB::Object::TYPE;

        let mut bytes = Self::serialized_prefix(o_type, content);
        bytes.extend_from_slice(content);
        let oid = Oid::for_serialized_bytes(&bytes);

        helper::<OB::Object>(self, &oid, &bytes).map_err(|e| StoreError(oid, e))?;

        Ok(oid)
    }

    fn serialized_prefix(o_type: &[u8], serialized: &[u8]) -> Vec<u8> {
        let size = serialized.len().to_string();

        let mut ser = Vec::with_capacity(o_type.len() + 1 + size.len() + 1);

        ser.extend(o_type);
        ser.push(b' ');
        ser.extend(size.as_bytes());
        ser.push(b'\0');

        ser
    }

    fn oid_path<O: Object>(&self, oid: &Oid<O>) -> PathBuf {
        let oid = oid.to_hex();
        let dir = self.path.join(&oid[0..2]);
        let name = &oid[2..];
        dir.join(name)
    }
}

impl Clone for Db {
    fn clone(&self) -> Self {
        Self::new(&self.path)
    }
}

pub type StoreResult<O> = Result<Oid<O>, StoreError<O>>;

#[derive(Debug, thiserror::Error, displaydoc::Display)]
/// Failed to store {0:?}
pub struct StoreError<O: Object>(Oid<O>, #[source] io::Error);

#[derive(Debug, thiserror::Error, displaydoc::Display)]
pub enum LoadError<O: Object + 'static> {
    /// Failed to load bytes of object {0:?}
    LoadBytes(#[from] LoadBytesError<O>),
    /// Failed to deserialize {0:?}
    Deserialize(Oid<O>, #[source] O::DeserializeError),
}

#[derive(Debug, thiserror::Error, displaydoc::Display)]
pub enum LoadBytesError<O: Object + 'static> {
    /// {0:?} not found in database
    NotFound(Oid<O>),
    /// Failed to open the file for {0:?} in the database
    Open(Oid<O>, #[source] io::Error),
    /// Failed to read the prefix from the file for {0:?} in the database
    ReadPrefix(Oid<O>, #[source] io::Error),
    /// Database entry for {0:?} is corrupt
    Corrupt(Oid<O>),
    /// Expected oid {oid:?} to have type {expected}, got {actual}
    WrongType {
        oid: Oid<O>,
        expected: BString,
        actual: BString,
    },
    /// Failed to parse bytes of length of {0:?} as utf-8
    ParseLenToBytes(Oid<O>, #[source] bstr::Utf8Error),
    /// Failed to parse length of {0:?}
    ParseLenToInt(Oid<O>, #[source] ParseIntError),
}
