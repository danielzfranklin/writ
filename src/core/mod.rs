pub mod db;
pub mod index;
pub mod locked_file;
pub mod refs;
pub mod repo;
pub mod stat;
pub mod status;
pub mod with_digest;
pub mod ws;

pub use db::{Db, Object, ObjectBuilder, Oid};
pub use index::{Index, IndexMut};
pub use locked_file::LockedFile;
pub use refs::Refs;
pub use repo::Repo;
pub use stat::Stat;
pub use status::{FileStatus, Status};
pub use with_digest::WithDigest;
pub use ws::Workspace;
pub use ws::WsPath;
