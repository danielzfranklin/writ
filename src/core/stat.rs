use std::{
    convert::TryInto,
    fs,
    os::linux::fs::MetadataExt,
    time::{Duration, SystemTime},
};
use tracing::warn;

use bstr::{BStr, ByteSlice};

#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub struct Stat {
    pub ctime: SystemTime,
    pub mtime: SystemTime,
    pub dev: u32,
    pub ino: u32,
    pub mode: Mode,
    pub uid: u32,
    pub gid: u32,
    pub size: u32,
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum Mode {
    Regular,
    Executable,
}

impl Stat {
    #[allow(
        clippy::cast_possible_truncation,
        clippy::cast_sign_loss,
        clippy::similar_names
    )]
    pub fn from(meta: &fs::Metadata) -> Self {
        Self {
            ctime: Self::systemtime_from_epoch(meta.st_ctime() as u32, meta.st_ctime_nsec() as u32),
            mtime: Self::systemtime_from_epoch(meta.st_mtime() as u32, meta.st_mtime_nsec() as u32),
            dev: meta.st_dev() as u32,
            ino: meta.st_ino() as u32,
            mode: Mode::from_u32(meta.st_mode()),
            uid: meta.st_uid(),
            gid: meta.st_gid(),
            size: meta.st_size() as u32,
        }
    }

    pub fn zeroed() -> Self {
        Self {
            ctime: SystemTime::UNIX_EPOCH,
            mtime: SystemTime::UNIX_EPOCH,
            dev: 0,
            ino: 0,
            mode: Mode::Regular,
            uid: 0,
            gid: 0,
            size: 0,
        }
    }

    pub fn ctime_epoch(&self) -> (u32, u32) {
        Self::systemtime_to_epoch(self.ctime)
    }

    pub fn mtime_epoch(&self) -> (u32, u32) {
        Self::systemtime_to_epoch(self.mtime)
    }

    fn systemtime_to_epoch(time: SystemTime) -> (u32, u32) {
        let dur = time
            .duration_since(SystemTime::UNIX_EPOCH)
            .expect("Not before epoch");

        let secs: u32 = dur.as_secs().try_into().expect("Time overflowed");

        (secs, dur.subsec_nanos())
    }

    pub(crate) fn systemtime_from_epoch(secs: u32, nanos: u32) -> SystemTime {
        SystemTime::UNIX_EPOCH + Duration::new(u64::from(secs), nanos)
    }
}

impl Mode {
    const EXECUTABLE: u32 = 0o10_07_55;
    const REGULAR: u32 = 0o10_06_44;

    const REGULAR_S: &'static [u8] = b"100644";
    const EXECUTABLE_S: &'static [u8] = b"100755";

    pub fn as_base8(self) -> &'static BStr {
        match self {
            Self::Regular => Self::REGULAR_S.as_bstr(),
            Self::Executable => Self::EXECUTABLE_S.as_bstr(),
        }
    }

    pub fn as_u32(self) -> u32 {
        match self {
            Self::Regular => Self::REGULAR,
            Self::Executable => Self::EXECUTABLE,
        }
    }

    pub fn from_u32(val: u32) -> Self {
        let is_executable = val & 0o111 != 0;
        if is_executable {
            Self::Executable
        } else {
            Self::Regular
        }
    }

    /// Unrecognized modes are considered [`Self::Regular`]
    pub fn from_base8(bytes: &BStr) -> Self {
        match bytes.as_bytes() {
            Self::REGULAR_S => Self::Regular,
            Self::EXECUTABLE_S => Self::Executable,
            _ => {
                warn!("Assuming unrecognized mode {} to be regular", bytes);
                Self::Regular
            }
        }
    }
}
