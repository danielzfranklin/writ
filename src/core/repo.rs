use std::{
    collections::BTreeMap,
    env, fmt, fs,
    io::{self},
    path::{Path, PathBuf},
};

use crate::core::{
    db::{self, object, tree, Blob, Commit, Tree},
    index::{
        self,
        entry::{self, Entry, StatusChatty},
    },
    refs,
    ws::{self, ListFilesError, ReadFileError, StatFileError},
    Db, FileStatus, Index, IndexMut, ObjectBuilder, Oid, Refs, Status, Workspace, WsPath,
};
use chrono::Local;
use tracing::{debug, instrument};

#[derive(Debug, Clone)]
pub struct Repo {
    git_dir: PathBuf,
    pub workspace: Workspace,
    pub db: Db,
    pub refs: Refs,
    pub index: Index,
}

impl Repo {
    #[instrument(err)]
    pub fn new(workspace: impl Into<PathBuf> + fmt::Debug) -> Result<Self, ReadError> {
        let workspace_dir = workspace.into();
        let workspace_dir = workspace_dir
            .canonicalize()
            .map_err(|e| ReadError::Io(workspace_dir, e))?;

        let git_dir = workspace_dir.join(".git");

        if !git_dir
            .try_exists()
            .map_err(|e| ReadError::Io(git_dir.clone(), e))?
        {
            return Err(ReadError::NotRepo(workspace_dir));
        }

        let workspace = Workspace::new(workspace_dir);
        let db = Db::new(&git_dir);
        let refs = Refs::new(&git_dir);
        let index = Index::load(&git_dir)?;

        Ok(Self {
            git_dir,
            workspace,
            db,
            refs,
            index,
        })
    }

    pub fn for_current_dir() -> Result<Self, ForCurrentDirError> {
        let dir = env::current_dir()?;
        Ok(Self::new(dir)?)
    }

    #[instrument(err)]
    pub fn init(workspace: impl Into<PathBuf> + fmt::Debug) -> Result<Self, InitError> {
        let workspace_dir = workspace.into();

        fs::create_dir_all(&workspace_dir)
            .map_err(|e| InitError::CreateWorkspace(workspace_dir.clone(), e))?;

        let git_dir = workspace_dir.join(".git");
        if git_dir
            .try_exists()
            .map_err(|e| InitError::Open(git_dir.clone(), e))?
        {
            return Err(InitError::Exists(git_dir));
        }

        for child in &["objects", "refs"] {
            let child = git_dir.join(child);
            fs::create_dir_all(&child).map_err(|e| InitError::Write(child, e))?;
        }

        let workspace = Workspace::new(workspace_dir);
        let db = Db::new(&git_dir);
        let refs = Refs::new(&git_dir);
        let index = Index::load(&git_dir)?;

        Ok(Self {
            git_dir,
            workspace,
            db,
            refs,
            index,
        })
    }

    #[instrument(err)]
    pub fn add<I, P>(&mut self, files: I) -> Result<(), AddError>
    where
        I: IntoIterator<Item = P> + fmt::Debug,
        P: AsRef<Path>,
    {
        let workspace = &self.workspace;
        let db = &self.db;
        self.index.reload()?;
        let mut index = self.index.modify()?;

        for file in workspace.find_files(files)? {
            let data = workspace.read_file(&file)?;
            let stat = workspace.stat(&file)?;

            let oid = db::blob::Builder::new(data).store(&db)?;
            let entry = Entry::new(file, oid, stat);

            debug!("Adding {:?}", entry);
            index.add(entry);
        }

        index.commit()?;

        Ok(())
    }

    #[instrument(err)]
    pub fn commit(
        &mut self,
        name: impl Into<String> + fmt::Debug,
        email: impl Into<String> + fmt::Debug,
        msg: impl Into<String> + fmt::Debug,
    ) -> Result<(), CommitError> {
        let mut msg = msg.into();
        if msg.is_empty() {
            return Err(CommitError::EmptyMessage);
        }
        if !msg.ends_with('\n') {
            msg.push('\n');
        }

        let name = name.into();
        let email = email.into();

        let db = &self.db;
        let refs = &self.refs;
        let index = &self.index;

        let entries = index.entries().map(|entry| db::tree::EntryBuilder {
            oid: entry.oid,
            path: entry.path.clone(),
            mode: entry.mode(),
        });

        let root = db::tree::Builder::new().entries(entries).store(&db)?;

        let parent = refs.head()?;
        let author = db::Author::new_local(name, email, Local::now());
        let commit = db::commit::Builder::new(parent, root, author, msg).store(db)?;
        refs.update_head(&commit)?;

        Ok(())
    }

    /// Unlike git, this lists files only. Children of untracked directories are
    /// reported instead of reporting the directory itself.
    #[instrument(err)]
    pub fn status(&mut self) -> Result<BTreeMap<WsPath, FileStatus>, StatusError> {
        let head = if let Some(head) = self.refs.head()? {
            let tree = self.db.load(head)?.tree;
            self.db.load_tree_files(&WsPath::root(), tree)?
        } else {
            BTreeMap::new()
        };

        self.index.reload()?;
        let mut index = self
            .index
            .modify()
            .map_err(|e| StatusError::UpdateIndex(e.into()))?;

        let work = &self.workspace;

        let mut ws_statuses = BTreeMap::new();
        let mut index_statuses = BTreeMap::new();

        for path in self.workspace.list_files()? {
            let ws_status = Self::workspace_status_of(work, &mut index, &path)?;
            let index_status = Self::index_status_of(&index, &head, &path)?;
            debug!("{path} in workspace, so ws: {ws_status:?}, idx: {index_status:?}");
            ws_statuses.insert(path.clone(), ws_status);
            index_statuses.insert(path, index_status);
        }

        for entry in index.entries() {
            if !ws_statuses.contains_key(&entry.path) {
                debug!("{} in idx but not ws, so ws: Status::Deleted", entry.path);
                ws_statuses.insert(entry.path.clone(), Status::Deleted);
            }
        }

        for (path, _file) in head {
            if !index.is_tracked_file(&path) {
                debug!("{path} in head but not idx, so idx: Status::Deleted",);
                index_statuses.insert(path, Status::Deleted);
            }
        }

        index
            .commit()
            .map_err(|e| StatusError::UpdateIndex(e.into()))?;

        let mut statuses = BTreeMap::new();

        for (path, ws_status) in ws_statuses {
            let index_status = index_statuses.remove(&path).unwrap_or(Status::Untracked);
            statuses.insert(
                path.clone(),
                FileStatus {
                    path,
                    workspace: ws_status,
                    index: index_status,
                },
            );
        }

        for (path, index_status) in index_statuses {
            statuses.insert(
                path.clone(),
                FileStatus {
                    path,
                    workspace: Status::Deleted,
                    index: index_status,
                },
            );
        }

        Ok(statuses)
    }

    pub fn workspace_status_of(
        work: &Workspace,
        index: &mut IndexMut,
        path: &WsPath,
    ) -> Result<Status, StatusError> {
        let status = if let Some(entry) = index.entry(path) {
            match entry.index_status_chatty(work)? {
                StatusChatty::Unmodified => Status::Unmodified,
                StatusChatty::UnmodifiedButNewStat(new_stat) => {
                    index.update_stat(path, new_stat).expect("Entry exists");
                    Status::Unmodified
                }
                StatusChatty::Modified => Status::Modified,
                StatusChatty::Deleted => Status::Deleted,
            }
        } else {
            Status::Untracked
        };
        Ok(status)
    }

    #[allow(clippy::option_if_let_else)]
    pub fn index_status_of(
        index: &Index,
        head: &BTreeMap<WsPath, tree::FileNode>,
        path: &WsPath,
    ) -> Result<Status, StatusError> {
        let index_entry = if let Some(index_entry) = index.entry(path) {
            index_entry
        } else {
            return Ok(Status::Untracked);
        };

        let status = if let Some(head_file) = head.get(path) {
            if head_file.mode == index_entry.mode() && head_file.oid == index_entry.oid {
                Status::Unmodified
            } else {
                Status::Modified
            }
        } else {
            Status::Added
        };

        Ok(status)
    }

    pub fn show_head(&mut self) -> eyre::Result<()> {
        let head = self.refs.head()?.ok_or_else(|| eyre::eyre!("No HEAD"))?;
        let commit = self.db.load::<Commit>(head)?;
        eprintln!("HEAD: {}\n", head);
        self.print_tree(commit.tree, 0)?;
        Ok(())
    }

    fn print_tree(&mut self, tree: Oid<Tree>, level: usize) -> eyre::Result<()> {
        let tree = self.db.load::<Tree>(tree)?;
        let level_prefix = " ".repeat(level * 4);
        for node in tree.direct_children() {
            match node {
                db::tree::Node::File(db::tree::FileNode { name, mode, oid }) => {
                    println!("{}{} {} ({:?})", level_prefix, oid, name, mode)
                }
                db::tree::Node::Tree { name, oid } => {
                    println!("{}{} {}/", level_prefix, oid, name);
                    self.print_tree(*oid, level + 1)?;
                }
            }
        }
        Ok(())
    }
}

#[derive(Debug, displaydoc::Display, thiserror::Error)]
// Failed in initialize repository
pub enum InitError {
    /// Directory {0:?} already exists
    Exists(PathBuf),
    /// Failed to create workspace directory {0:?}
    CreateWorkspace(PathBuf, #[source] io::Error),
    /// Failed to open directory {0:?} to initialize
    Open(PathBuf, #[source] io::Error),
    /// Failed to populate {0:?}
    Write(PathBuf, #[source] io::Error),
    /// Failed to open index
    OpenIndex(#[from] index::LoadError),
}

#[derive(Debug, thiserror::Error, displaydoc::Display)]
/// Failed to read a directory as a git repository.
pub enum ReadError {
    /// The directory {0:?} is not a git repository
    NotRepo(PathBuf),
    /// IO error while checking if directory {0:?} is a git repository
    Io(PathBuf, #[source] io::Error),
    /// Failed to open index
    OpenIndex(#[from] index::LoadError),
}

#[derive(Debug, thiserror::Error, displaydoc::Display)]
/// Failed to read the current directory as a git repository.
pub enum ForCurrentDirError {
    /// Failed to get the current directory
    GetCurrentDir(#[from] io::Error),
    /// Failed to read git repository
    ReadError(#[from] ReadError),
}

#[derive(Debug, thiserror::Error, displaydoc::Display)]
pub enum AddError {
    /// Failed to reload index
    ReloadIndex(#[from] index::LoadError),
    /// Failed to open index of modifications
    OpenIndex(#[from] index::OpenForModificationsError),
    /// Failed to find files provided in repository
    FindFiles(#[from] ws::ListFilesError),
    /// Failed to stat file
    Stat(#[from] StatFileError),
    /// Failed to read file
    Read(#[from] ReadFileError),
    /// Failed to store file
    StoreBlob(#[from] db::StoreError<Blob>),
    /// Failed to commit changes to index
    CommitIndex(#[from] index::CommitError),
}

#[derive(Debug, thiserror::Error, displaydoc::Display)]
pub enum CommitError {
    /// Empty commit message
    EmptyMessage,
    /// Failed to load index
    LoadIndex(#[from] index::OpenForModificationsError),
    /// Failed to store tree
    StoreTree(#[from] db::StoreError<Tree>),
    /// Failed to store commit
    StoreCommit(#[from] db::StoreError<Commit>),
    /// Failed to read ref
    ReadRef(#[from] refs::ReadError),
    /// Failed to parse oid of parent commit
    ParseParentOid(#[from] object::ParseOidError),
    /// Failed to update ref
    UpdateRef(#[from] refs::UpdateError),
}

#[derive(Debug, thiserror::Error, displaydoc::Display)]
pub enum StatusError {
    /// Failed to reload index
    ReloadIndex(#[from] index::LoadError),
    /// Failed to get head oid
    GetHeadOid(#[from] refs::ReadError),
    /// Failed to load head commit
    LoadHeadCommit(#[from] db::LoadError<db::Commit>),
    /// Failed to load of tree from head
    LoadHeadTree(#[from] db::LoadError<db::Tree>),
    /// Failed to list files
    ListFiles(#[from] ListFilesError),
    /// Failed to check if file unchanged
    IsUnchanged(#[from] entry::IsUnchangedError),
    /// Failed to update index with new stat
    UpdateIndex(#[from] index::ModifyError),
}
