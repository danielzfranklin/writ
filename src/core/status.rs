use crate::core::WsPath;

#[derive(Debug, Clone, Eq, PartialEq)]
#[allow(clippy::module_name_repetitions)]
pub struct FileStatus {
    pub path: WsPath,
    pub index: Status,
    pub workspace: Status,
}

#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub enum Status {
    Untracked,
    Modified,
    Unmodified,
    Deleted,
    Added,
}

impl Status {
    pub fn name(self) -> &'static str {
        match self {
            Status::Untracked => "untracked",
            Status::Modified => "modified",
            Status::Unmodified => "unmodified",
            Status::Deleted => "deleted",
            Status::Added => "added",
        }
    }
}
