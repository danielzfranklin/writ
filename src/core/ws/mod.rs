pub mod path;
pub use path::WsPath;

use crate::core::Stat;

use bstr::BString;
use std::{
    fmt, fs, io,
    os::unix::ffi::OsStrExt,
    path::{Path, PathBuf},
};
use tracing::instrument;

#[derive(Debug, Clone)]
pub struct Workspace {
    path: PathBuf,
}

impl Workspace {
    const IGNORE: &'static [&'static [u8]] = &[b".git"];

    pub fn new<P: Into<PathBuf>>(path: P) -> Self {
        Self { path: path.into() }
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    #[instrument(err)]
    pub fn find_files<I, P>(&self, paths: I) -> Result<Vec<WsPath>, ListFilesError>
    where
        I: IntoIterator<Item = P> + fmt::Debug,
        P: AsRef<Path>,
    {
        let mut files = Vec::new();

        for rel_path in paths {
            let rel_path = rel_path.as_ref();
            let abs_path = self
                .path
                .join(rel_path)
                .canonicalize()
                .map_err(|e| ListFilesError::Canonicalize(rel_path.to_owned(), e))?;
            self.list_files_in(&abs_path, &mut files)?;
        }

        Ok(files)
    }

    #[instrument(err)]
    pub fn list_files(&self) -> Result<Vec<WsPath>, ListFilesError> {
        let mut files = Vec::new();
        self.list_files_in(&self.path, &mut files)?;
        Ok(files)
    }

    fn list_files_in(
        &self,
        abs_path: &Path,
        files: &mut Vec<WsPath>,
    ) -> Result<(), ListFilesError> {
        // we re-compute this to canonicalize
        let rel_path = abs_path
            .strip_prefix(&self.path)
            .map_err(|_| ListFilesError::OutsideOfWorkspace(abs_path.to_owned()))?;

        let meta = abs_path
            .metadata()
            .map_err(|e| ListFilesError::GetMetadata(abs_path.to_owned(), e))?;

        if Self::is_ignored(&rel_path) {
            return Ok(());
        }

        if meta.is_dir() {
            let children = abs_path
                .read_dir()
                .map_err(|e| ListFilesError::ReadDir(abs_path.to_owned(), e))?
                .map(|entry| {
                    let entry =
                        entry.map_err(|e| ListFilesError::ReadDirEntry(abs_path.to_owned(), e))?;
                    Ok(rel_path.join(entry.file_name()))
                })
                .collect::<Result<Vec<_>, ListFilesError>>()?;

            files.extend(self.find_files(children)?);
        } else if meta.is_file() {
            files.push(WsPath::new_unchecked(rel_path));
        } else {
            return Err(ListFilesError::InvalidFileType(rel_path.into()));
        }

        Ok(())
    }

    fn is_ignored(rel_path: &Path) -> bool {
        Self::IGNORE
            .iter()
            .any(|&ignored| rel_path.as_os_str().as_bytes() == ignored)
    }

    pub fn read_file(&self, path: &WsPath) -> Result<BString, ReadFileError> {
        let bytes = fs::read(path.to_absolute(self)).map_err(|e| ReadFileError(path.clone(), e))?;
        Ok(bytes.into())
    }

    pub fn stat(&self, path: &WsPath) -> Result<Stat, StatFileError> {
        self.path
            .join(path)
            .metadata()
            .map(|m| Stat::from(&m))
            .map_err(|e| StatFileError(path.clone(), e))
    }
}

#[derive(Debug, displaydoc::Display, thiserror::Error)]
/// Failed to stat file {0:?}
pub struct StatFileError(WsPath, io::Error);

impl StatFileError {
    pub(crate) fn is_not_found(&self) -> bool {
        self.1.kind() == io::ErrorKind::NotFound
    }
}

#[derive(Debug, displaydoc::Display, thiserror::Error)]
/// Failed to read file {0:?}
pub struct ReadFileError(WsPath, io::Error);

#[derive(Debug, displaydoc::Display, thiserror::Error)]
pub enum ListFilesError {
    /// {0:?} is neither a file nor a directory.
    InvalidFileType(PathBuf),
    /// Path {0:?} is outside the workspace
    OutsideOfWorkspace(PathBuf),
    /// Failed to canonicalize path {0:?}
    Canonicalize(PathBuf, #[source] io::Error),
    /// Failed to get metadata of {0:?}
    GetMetadata(PathBuf, #[source] io::Error),
    /// Failed to read directory {0:?}
    ReadDir(PathBuf, #[source] io::Error),
    /// Failed to read entry of directory {0:?}
    ReadDirEntry(PathBuf, #[source] io::Error),
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_support::init;
    use insta::assert_debug_snapshot;
    use pretty_assertions::assert_eq;
    use tempfile::tempdir;

    #[test]
    fn excludes_git_dir() -> eyre::Result<()> {
        init();

        let dir = tempdir()?;
        let dir = dir.path();

        fs::create_dir(dir.join(".git"))?;
        fs::write(dir.join(".git/a"), "foo")?;
        fs::write(dir.join("b"), "foo")?;

        let workspace = Workspace::new(dir);
        let actual = workspace.find_files(vec!["."])?;

        let expected = vec![WsPath::new_unchecked("b")];

        assert_eq!(expected, actual);

        Ok(())
    }

    #[test]
    fn find_files() -> eyre::Result<()> {
        init();

        let dir = tempdir()?;
        let dir = dir.path();

        fs::create_dir_all(dir.join("dir_1/dir_a"))?;
        fs::create_dir_all(dir.join("dir_1/dir_b"))?;

        fs::write(dir.join("a"), "")?;
        fs::write(dir.join("b"), "")?;

        fs::write(dir.join("dir_1/c"), "")?;
        fs::write(dir.join("dir_1/d"), "")?;

        fs::write(dir.join("dir_1/dir_a/e"), "")?;

        let workspace = Workspace::new(dir);

        let mut files = workspace.find_files(vec!["a"])?;
        files.sort();
        assert_debug_snapshot!("a", files);

        let mut files = workspace.find_files(vec!["."])?;
        files.sort();
        assert_debug_snapshot!("dot", files);

        let mut files = workspace.find_files(vec!["dir_1"])?;
        files.sort();
        assert_debug_snapshot!("dir_1", files);

        Ok(())
    }
}
