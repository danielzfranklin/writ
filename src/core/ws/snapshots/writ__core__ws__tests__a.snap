---
source: src/core/ws/mod.rs
expression: files

---
[
    WsPath(
        "a",
    ),
]
